django/db/models/fields/tuple_lookups.py

```diff
@@ -0,0 +1,222 @@
+
+
+class TupleLessThan(TupleLookupMixin, LessThan):
+    def as_sql(self, compiler, connection):
+        from django.db.models.sql.where import AND, OR, WhereNode
+
+        # e.g.: (a, b, c) < (x, y, z) as SQL:
+        # WHERE a < x OR (a = x AND (b < y OR (b = y AND c < z)))
+        cols = self.lhs.get_cols()
+        lookups = itertools.cycle([LessThan, Exact])  # <, =, <, =, ...
+        connectors = itertools.cycle([OR, AND])  # OR, AND, OR, AND, ...
+        cols_list = [col for col in cols for _ in range(2)]
+        vals_list = [val for val in self.rhs for _ in range(2)]
+        cols_iter = iter(cols_list[:-1])  # a, a, b, b, c
+        vals_iter = iter(vals_list[:-1])  # x, x, y, y, z
+        col, val = next(cols_iter), next(vals_iter)
+        lookup, connector = next(lookups), next(connectors)
+        root = node = WhereNode([lookup(col, val)], connector=connector)
+
+        for col, val in zip(cols_iter, vals_iter):
+            lookup, connector = next(lookups), next(connectors)
+            child = WhereNode([lookup(col, val)], connector=connector)
+            node.children.append(child)
+            node = child
+
+        return root.as_sql(compiler, connection)
+
+
+class TupleLessThanOrEqual(TupleLookupMixin, LessThanOrEqual):
+    def as_sql(self, compiler, connection):
+        from django.db.models.sql.where import AND, OR, WhereNode
+
+        # e.g.: (a, b, c) <= (x, y, z) as SQL:
+        # WHERE a < x OR (a = x AND (b < y OR (b = y AND (c < z OR c = z))))
+        cols = self.lhs.get_cols()
+        lookups = itertools.cycle([LessThan, Exact])  # <, =, <, =, ...
+        connectors = itertools.cycle([OR, AND])  # OR, AND, OR, AND, ...
+        cols_list = [col for col in cols for _ in range(2)]
+        vals_list = [val for val in self.rhs for _ in range(2)]
+        cols_iter = iter(cols_list)  # a, a, b, b, c, c
+        vals_iter = iter(vals_list)  # x, x, y, y, z, z
+        col, val = next(cols_iter), next(vals_iter)
+        lookup, connector = next(lookups), next(connectors)
+        root = node = WhereNode([lookup(col, val)], connector=connector)
+
+        for col, val in zip(cols_iter, vals_iter):
+            lookup, connector = next(lookups), next(connectors)
+            child = WhereNode([lookup(col, val)], connector=connector)
+            node.children.append(child)
+            node = child
+
+        return root.as_sql(compiler, connection)
+
+
+class TupleIn(TupleLookupMixin, In):
+    def check_tuple_lookup(self):
+        assert isinstance(self.lhs, Cols)
+        self.check_rhs_is_tuple_or_list()
+        self.check_rhs_is_collection_of_tuples_or_lists()
+        self.check_rhs_elements_length_equals_lhs_length()
+
+    def as_sql(self, compiler, connection):
+        from django.db.models.sql.where import AND, OR, WhereNode
+
+        if not self.rhs:
+            raise EmptyResultSet
+
+        # e.g.: (a, b, c) in [(x1, y1, z1), (x2, y2, z2)] as SQL:
```

Comment:
This one seems to work even on [Oracle](https://dbfiddle.uk/C2GjO1o5)?

Reply:
I believe this is the only one that will work on Oracle.